### DISCUSSION TITLE
Legacy version: Pulled docker hub images?

### DISCUSSION THREAD

**Discussion by p-hebert** [apache/airflow#42861] [Q&A] [answered]
Hello wonderful Airflow maintainers 👋 🙂 

_You either die a hero, or live to see yourself maintain legacy code within your codebase._
_I would never have thought to see this day._ 😭 😆 

We are currently running an older legacy version of Airflow (won't specifically say which, for security reasons), and we would like to upgrade to the same exact semantic version but with python3.9 instead of python3.8.
Unfortunately, from what I see in the list of tags on Docker Hub, there are no stable releases for a good few minor versions; all of them are release candidates (rc1, rc2, rc3, ...) up until 2.7.0, and we don't necessarily have the manpower in the short term to do such a significant upgrade with potentially breaking changes and deprecations.

Is there somewhere I could pull the stable docker images for legacy versions?
Alternatively, is there a _real easy_ way to pull the repository and build the image from scratch?

Thanks a lot! 🙏 

Have a great day!

**Comment by potiuk** [comment, marked as answer]
No. All versions are in dockerhub (for all supported python versions that particular version was supporting). I think you likely have not seen that results are paged and the right one might be on paged results - or you could filter dockerhub by exact versions (say `2.6.2-python3.8`:

<img width="799" alt="Screenshot 2024-10-09 at 15 00 43" src="https://github.com/user-attachments/assets/855ee822-5af0-4a8f-81c4-da053e12a6a2">

**Reply by p-hebert** [reply]
Hi Jared! Yes, no I'm aware that there's pagination 😅 
I'm looking for python3.9, not python3.8, and if I filter using python3.9 in the tag search, the earliest non release-candidate version is 2.7.0-python3.9. So yes, a few minor version ahead of our currently in use version.
So from what I gather from your explanation, there aren't any stable release python3.9 version below 2.7.0?

**Reply by andresitorusz** [reply]
I just checked the Airflow Docker Hub and can tell that the earliest stable release of Python 3.9 on the Airflow image was [2.1.2-python3.9](https://hub.docker.com/layers/apache/airflow/2.1.2-python3.9/images/sha256-6b440a1c77c079f93d21f4028608278d78f9ef60a3f6cee1efbeb2728039809b).

**Reply by p-hebert** [reply]
Aaaaaaahh OK, I found the issue; the sorting is by newest to oldest and I didn't realize that python3.9 versions would be newer than a lot of minor version upgrades. I was so confused XD

Thank you! I have found the version I was looking for by searching for the minor version and scrolling through; My badd!

**Reply by potiuk** [reply]
Yeah. Things might get confusing :)
